File: backend/app/schemas/delivery_otp.py
"""Delivery OTP API schemas."""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import DeliveryOtpStatus


class DeliveryOtpVerifyRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    code: str = Field(min_length=6, max_length=6, pattern=r"^\d{6}$")
    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)


class DeliveryVerificationStatusResponse(BaseModel):
    order_id: UUID
    status: DeliveryOtpStatus
    generated_at: datetime | None = None
    expires_at: datetime | None = None
    verified_at: datetime | None = None
    delivery_agent_user_id: UUID | None = None
    verification_latitude: Decimal | None = None
    verification_longitude: Decimal | None = None
    failed_attempts: int = 0
    is_verified: bool = False
    otp_available: bool = False


class CustomerDeliveryOtpResponse(BaseModel):
    order_id: UUID
    otp_code: str
    expires_at: datetime
    status: DeliveryOtpStatus


class DeliveryVerifyCompleteResponse(BaseModel):
    order_id: UUID
    status: str
    verified_at: datetime
    delivery_agent_user_id: UUID

File: frontend/services/delivery-otp.ts
import { api, type ApiEnvelope } from '@/lib/api';

export type DeliveryOtpStatus = 'active' | 'verified' | 'expired' | 'locked';

export interface DeliveryVerificationStatus {
  order_id: string;
  status: DeliveryOtpStatus;
  generated_at: string | null;
  expires_at: string | null;
  verified_at: string | null;
  delivery_agent_user_id: string | null;
  verification_latitude: string | null;
  verification_longitude: string | null;
  failed_attempts: number;
  is_verified: boolean;
  otp_available: boolean;
}

export interface CustomerDeliveryOtp {
  order_id: string;
  otp_code: string;
  expires_at: string;
  status: DeliveryOtpStatus;
}

export interface DeliveryVerifyResult {
  order_id: string;
  status: string;
  verified_at: string;
  delivery_agent_user_id: string;
}

export async function getCustomerDeliveryOtp(orderId: string): Promise<CustomerDeliveryOtp> {
  const { data } = await api.get<ApiEnvelope<CustomerDeliveryOtp>>(`/orders/${orderId}/delivery-otp`);
  return data.data;
}

export async function getPartnerDeliveryVerification(orderId: string): Promise<DeliveryVerificationStatus | null> {
  const { data } = await api.get<ApiEnvelope<DeliveryVerificationStatus | null>>(
    `/partner/orders/${orderId}/delivery-verification`,
  );
  return data.data;
}

export async function verifyDeliveryOtp(
  orderId: string,
  body: { code: string; latitude?: number; longitude?: number },
): Promise<DeliveryVerifyResult> {
  const { data } = await api.post<ApiEnvelope<DeliveryVerifyResult>>(
    `/partner/orders/${orderId}/delivery/verify`,
    body,
  );
  return data.data;
}
